lib/stellar/sep10.rb

```diff
@@ -0,0 +1,353 @@
+      # important when verifying signers of a challenge transaction that we only
+      # verify and return client signers. If an account has the server as a
+      # signer the server should not play a part in the authentication of the
+      # client.
+      client_signers = Array.new
+      signers.each do |signer|
+        if signer.address != server.address
+          client_signers.push(signer)
+        end
+      end
+
+      signers_found = verify_transaction_signatures(
+        transaction_envelope: te, signers: client_signers
+      )
+
+      # Confirm we matched signatures to the client signers.
+      if signers_found.empty?
+        raise InvalidSep10ChallengeError.new("Transaction not signed by any client signer.")
+      end
+
+      # Confirm all signatures were consumed by a signer.
+      if signers_found.length != te.signatures.length - 1
+        raise InvalidSep10ChallengeError.new("Transaction has unrecognized signatures.")
+      end
+
+      return signers_found
+
+    end
+
+    Contract(C::KeywordArgs[
+      challenge_transaction: String,
+      server: Stellar::KeyPair,
+      threshold: Integer,
+      signers: ArrayOf[Stellar::AccountSigner],
+    ] => C::ArrayOf[Stellar::AccountSigner])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction 
+    # are accounted for and that the signatures meet a threshold on an account. A 
+    # transaction is verified if it is signed by the server account, and all other 
+    # signatures match a signer that has been provided as an argument, and those 
+    # signatures meet a threshold on the account.
+    #
+    # @param challenge_transaction [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    # @param threshold [Integer] The medThreshold on the client account.
+    # @param signers [ArrayOf[Stellar::AccountSigner]]The signers of client account.
+    #
+    # @return [ArrayOf[Stellar::AccountSigner]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #   - The transaction is invalid according to Stellar::SEP10.read_challenge_transaction.
+    #   - One or more signatures in the transaction are not identifiable as the server 
+    #     account or one of the signers provided in the arguments.
+    #   - The signatures are all valid but do not meet the threshold.
+    def self.verify_challenge_transaction_threshold(
+      challenge_transaction:,
+      server:,
+      threshold:,
+      signers:
+    )
+      signers_found = verify_challenge_transaction_signers(
+        challenge: challenge_transaction, 
+        server: server, 
+        signers: signers
+      )
+  
+      weight = 0
+      signers_found.each do |s|
+        weight += s.weight
+      end
+
+      if weight < threshold
+        raise InvalidSep10ChallengeError.new(
+          "signers with weight %d do not meet threshold %d." % [weight, threshold]
```

Comment:
Highly recommend using interpolation for strings instead of the sprintf syntax, as interpolation is favored by the Ruby community.
```suggestion
          "signers with weight #{weight} do not meet threshold #{threshold}."
```
There are some other places in the code being added that you can use interpolation too.